csrc/scheduler/reduction.cpp

```diff
@@ -507,40 +507,253 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
   return rparams;
 }
 
-std::shared_ptr<ReductionParams> outerReductionHeuristic(
+struct OuterReduHeuristicParas {
+  OuterReduHeuristicParas(
+      int64_t total_iteration_numel,
+      int64_t total_reduction_numel)
+      : total_iteration_numel(total_iteration_numel),
+        total_reduction_numel(total_reduction_numel) {}
+  // iteration dim paras
+  // iteration elements = iter_unroll * bdimx * gidim
+  int64_t iter_unroll_factor = 1;
+  int64_t bdimx = 1;
+  int64_t gidim = 1;
+  // reduction dim paras
+  // reduction elments = redu_unroll * bdimy * grdim * redu_serial
+  int64_t redu_unroll_factor = 1;
+  int64_t bdimy = 1;
+  int64_t grdim = 1;
+  int64_t redu_serial = 1;
+
+  // iteration and reduction dim elements
+  int64_t total_iteration_numel = -1;
+  int64_t total_reduction_numel = -1;
+
+  // Helper to figure out how much is left in the iter or reduction dim
+  int64_t iDimAvail() const {
+    return ceilDiv(total_iteration_numel, gidim * bdimx * iter_unroll_factor);
+  }
+  int64_t rDimAvail() const {
+    return ceilDiv(total_reduction_numel, grdim * bdimy * redu_unroll_factor);
+  };
+
+  std::string toString() const {
+    std::stringstream ss;
+    ss << "\n===== Outer Reduction Stats ========\n"
+       << "total_reduction_numel: " << total_reduction_numel << "\n"
+       << "total_iteration_numel: " << total_iteration_numel << "\n"
+       << "vectorize_factor: " << iter_unroll_factor << "\n"
+       << "redu_unroll_factor: " << redu_unroll_factor << "\n"
+       << "grid(" << gidim << ", " << grdim << ", 1)"
+       << "\n"
+       << "block(" << bdimx << ", " << bdimy << ", 1)" << std::endl;
+    return ss.str();
+  }
+
+  // compare block reduction with grid reduction
+  bool isBetterThan(const OuterReduHeuristicParas& grid_hp, int64_t sm_count)
+      const {
+    NVF_ERROR(
+        grdim == 1,
+        "Only support compare block reduction heuristic with grid reduction not vice versa");
+
+    // use block reduction if its SM usage >= 90% and its iter_unroll_factor is
```

Comment:
Please leave some more detailed background on why these conditions should be used. For example, if this was the first time I saw this logic, I'd wonder why the SM efficiency and unrolling factor of the block reduction were considered.

Reply:
revised comments:
```
  // use block reduction if its SM usage >= 90% and its iter_unroll_factor is
  // equal or larger than grid reduction. These two conditions ensure high SM
  // usage and efficient global memory access. The corresponding block reduction
  // avoids the overhead of inter-block data exchange through global memory.
  // It is faster than grid reduction even not all SMs are used.
  // TODO: if we know the fusion is memory bound (e.g. pure reduction), we can
  // use a lower threshold. For computation bound (e.g. gelu bwd), relaxing
  // the threshold leads to regression.
```